### DISCUSSION TITLE
Input for GCNConv layer (torch.sparse coo vs edge index and edge attr tensors)

### DISCUSSION THREAD

**Discussion by charleneolive** [pyg-team/pytorch_geometric#8588] [Q&A]
Hello,

I am slightly confused by the behaviour of GCNConv.
When I used the input as a torch.sparse_coo tensor vs having the edge_index and edge_attr placed as separate inputs, I get a different outcome from the gcnconv operation. Debugging it made me conclude that it was primarily due to the coalesce function, which has been applied to the torch.sparse_coo tensor but not for the scenario where I used edge_index and edge_attr as inputs. In particular, this happened during the add_self_loops_fn for the torch_sparse_coo_tensor, under to_torch_coo_tensor. 

I was trying to debug this as I was training a GCN model where the performance decreases when I used inputs as edge_index and edge_attr, instead of using as a sparse tensor. 

For a reproducible scenario, please refer to the collab link here (under Differences in output):
https://colab.research.google.com/drive/144kSk-OfSwZHChfnSHbpmB-iV_4ssCZY?usp=sharing

This is based on the example here: https://pytorch-geometric.readthedocs.io/en/latest/get_started/introduction.html

Would really appreciate if someone can enlighten me on the following
1. What is the purpose of coalesce and why is it not used for the other type of input? 
2.  How would it affect the performance when training the GCN?
3. What should be the appropriate inputs to the GCNConv? Can I use torch.sparse_coo_tensor? How can I get the same outcomes regardless of the inputs I used?

**Comment by rusty1s** [comment]
Note that for non-symmetric sparse tensors, you need to input the transposed adjacency matrix:
```python
x_out = conv1(x, ori_edge_index, edge_weight)
tensor_adjacency = to_torch_coo_tensor(data.edge_index, edge_weight)
x_out2 = conv1(x, tensor_adjacency.t())
print(torch.allclose(x_out, x_out2, atol=1e-4))
```
works for me.

**Comment by charleneolive** [comment]
Thank you so much for your quick reply!
I tried it out and met with an error
<img width="1256" alt="image" src="https://github.com/pyg-team/pytorch_geometric/assets/35939804/38ee0cc2-7e83-4024-8507-f8da12d47dbf">
Upon adding coalesce, the error is resolved:
<img width="982" alt="image" src="https://github.com/pyg-team/pytorch_geometric/assets/35939804/04adda4b-3792-4ff3-9d8e-52778127a41b">
Could I understand a bit more about why coalesce is required for the adjacency matrix?

**Reply by EdisonLeeeee** [reply]
Hi @charleneolive 

Coalescing is used to eliminate duplicate coordinates in the indices of a sparse adjacency matrix, which is necessary when accessing the `indices()` and `values()`.

FYI: https://pytorch.org/docs/stable/sparse.html

**Comment by charleneolive** [comment]
Hello @EdisonLeeeee,
Thank you so much. It makes sense now.

Could I ask if there are any considerations otherwise? Because when I tried it on my dataset, I still face a problem where they are not similar:
<img width="901" alt="image" src="https://github.com/pyg-team/pytorch_geometric/assets/35939804/254abcd8-a69c-4bbf-b377-2fd9226e6b61">

Unfortunately, I cannot share the dataset due to data privacy considerations. Perhaps if there's any advice what should we look out for to make sure the the values are similar? Thank you so much!

**Reply by EdisonLeeeee** [reply]
Weird. Can you check if all the values in `edge_weight > 0`?

**Reply by charleneolive** [reply]
Hello @EdisonLeeeee ,
Sorry for my late reply. 
Yes they all are. 

<img width="812" alt="image" src="https://github.com/pyg-team/pytorch_geometric/assets/35939804/9c0cd4e9-01ba-47d6-a1e0-701780bcfdd2">

**Reply by rusty1s** [reply]
On which PyG version are you running? It may be that edge weights are not supported yet for your used version.

**Reply by charleneolive** [reply]
Hello @rusty1s 

Thank you so much, I will check again. Can I ask from which pyG version are the edge weights and indices supported? I would check when I get the chance again.

**Reply by charleneolive** [reply]
Hello @rusty1s , 

I have checked and my pyG version is 2.3.1. Based on the documentation from 2.3.1, edge weights and edge indices are supported: https://pytorch-geometric.readthedocs.io/en/2.3.1/generated/torch_geometric.nn.conv.GCNConv.html?highlight=gcnconv#torch_geometric.nn.conv.GCNConv. 

For the adjacency matrix, it wasn't mentioned explicitly, but it was also able to get accepted, and under the gcn_norm  in GCNConv, it passes through "if is_torch_sparse_tensor(edge_index)"

Currently, the inputs of edge weights and edge indices show very poor convergence for the training dataset and a poor performance overall for the test dataset,  while for the adjacency matrix, it shows much better convergence.

**Reply by rusty1s** [reply]
Just confirmed that it works fine for me on 2.3.1 as well.

> Currently, the inputs of edge weights and edge indices show very poor convergence for the training dataset and a poor performance overall for the test dataset, while for the adjacency matrix, it shows much better convergence.

That's interesting, but I don't have a good explanation for this yet. If you ever manage to come up with a reproducible example, I am very curious to look into this.

**Reply by charleneolive** [reply]
I realised that the difference is caused by self loops. If I pass add_self_loops = False into GCNConv, I get the same outputs regardless of what type of input I used. I have also validated this by comparing the edge index and weights before and after adding self loops. For the adjacency matrix, the function used for adding self loops is add_self_loops_fn while for the inputs of edge_weight and edge_index, the function is add_remaining_self_loops

**Reply by charleneolive** [reply]
To add on to my previous comment, when I retrained my model again with `add_self_loops=False` for both using adjacency matrix and edge_index and edge_weights, I got the same result.

**Reply by charleneolive** [reply]
It's alright! I think I got it.
As for the adjacency matrix the repeated self-loops are not removed, this would cause two self-loops, as self-loops are already present in the adjacency matrix. For this dataset, maybe a stronger emphasis on the features of the node itself is required, and maybe works better in this scenario (please correct me if my understanding is wrong).
